ArduPlane/fence.cpp

```diff
@@ -0,0 +1,132 @@
+#include "Plane.h"
+
+// Code to integrate AC_Fence library with main ArduPlane code
+
+#if AC_FENCE == ENABLED
+
+// fence_check - ask fence library to check for breaches and initiate the response
+// called at 1Hz
+void Plane::fence_check()
+{
+    const uint8_t orig_breaches = fence.get_breaches();
+
+    // check for new breaches; new_breaches is bitmask of fence types breached
+    const uint8_t new_breaches = fence.check();
+
+    // we still don't do anything when disarmed, but we do check for fence breaches.
+    // fence pre-arm check actually checks if any fence has been breached
+    // that's not ever going to be true if we don't call check on AP_Fence while disarmed
+    if (!arming.is_armed())
+    {
+        return;
+    }
+
+    if (!fence.enabled())
+    {
+        // Switch back to the chosen control mode if still in
+        // GUIDED to the return point
+        if ((fence.get_action() == FENCE_ACTION_GUIDED || fence.get_action() == FENCE_ACTION_GUIDED_THR_PASS || fence.get_action() == FENCE_ACTION_RTL) &&
+            (control_mode == &mode_guided || control_mode == &mode_avoidADSB))
+        {
+            set_mode(*previous_mode, ModeReason::GCS_COMMAND);
+        }
+    }
+
+    // Never trigger a fence breach in the final stage of landing
+    if (landing.is_expecting_impact())
+    {
+        return;
+    }
+
+    if (new_breaches)
+    {
+        // if the user wants some kind of response and motors are armed
+        const uint8_t fence_act = fence.get_action();
+        switch (fence_act)
+        {
+        case AC_FENCE_ACTION_REPORT_ONLY:
+            break;
+        case AC_FENCE_ACTION_GUIDED:
+        case AC_FENCE_ACTION_GUIDED_THROTTLE_PASS:
+        case AC_FENCE_ACTION_RTL_AND_LAND:
+            // make sure we don't auto trim the surfaces on this mode change
+            int8_t saved_auto_trim = g.auto_trim;
+            g.auto_trim.set(0);
+            if (fence_act == AC_FENCE_ACTION_RTL_AND_LAND) {
+                set_mode(mode_rtl, ModeReason::FENCE_BREACHED);
+            } else {
+                set_mode(mode_guided, ModeReason::FENCE_BREACHED);
+            }
+            g.auto_trim.set(saved_auto_trim);
+
+            if (fence.get_return_rally() != 0 || fence_act == AC_FENCE_ACTION_RTL_AND_LAND) {
+                guided_WP_loc = rally.calc_best_rally_or_home_location(current_loc, get_RTL_altitude());
+            } else {
+                //return to fence return point, not a rally point
+                guided_WP_loc = {};
+                if (fence.get_return_altitude() > 0) {
```

Comment:
I wonder why we're not using this in the "true" clause in the if statement here.

Reply:
Again, another case of carrying across existing implementation.

`fence.get_return_altitude()` uses the fence parameter for the fence return point altitude, which can be specified separately to the RTL altitude, which is used in the "true" clause. Either side of the clause could be changed, but that's a change to existing behaviour.